lib-backend-wasm/src/func.rs

```diff
@@ -1331,15 +1595,66 @@ fn encode_appl<H: HeapManager>(
                     )),
                 wasmgen::TableIdx { idx: 0 },
             );
-        }
+        });
+    }
+}
 
-        // fetch return values from the location prescribed by the calling convention back to the stack
-        encode_post_appl_calling_conv(
-            Some(ir::VarType::Any),
+// Trampolining implemented using loops. Keep making making indirect calls
+// until [i32(is_tail == 1)] -> [i32(is_tail == 0)] implying no further tail calls.
+// Pop result values from unprotected stack that were stored by the tail call.
+// net wasm stack: [i32(is_tail)] -> [ir_return_type]
+fn encode_trampoline<H: HeapManager>(
+    return_type: Option<ir::VarType>,
+    ctx: EncodeContext<H>,
+    mutctx: &mut MutContext,
+    expr_builder: &mut wasmgen::ExprBuilder,
+) {
+    // If [i32(is_tail == 1)] run tail_calling conv again. If [i32(is_tail == 0)]
+    // then run post_appl_calling_conv already as no further tail calls.
+    mutctx.with_unused_landing(|mutctx| {
+        multi_value_polyfill::if_(
+            encode_opt_vartype(return_type),
             ctx.options.wasm_multi_value,
-            ctx.stackptr,
-            mutctx.scratch_mut(),
+            mutctx,
             expr_builder,
+            |mutctx, expr_builder| {
+                // net wasm stack: [i32(is_tail == 1)] -> [i32(is_tail == 1)] or
+                // net wasm stack: [i32(is_tail == 1)] -> [i32(is_tail === 0)]
+                encode_tail_calling_conv(ctx, mutctx, expr_builder);
+                expr_builder.if_(&[]);
+                {
+                    // Keep looping until the i32 value on the stack is 0.
+                    expr_builder.loop_(&[]);
+                    {
+                        encode_tail_calling_conv(ctx, mutctx, expr_builder);
+                        expr_builder.br_if(0);
+                    }
+                    expr_builder.end();
+                }
+                expr_builder.end();
+
+                // Now that no other tail calls need to be run, we can retreive
+                // the result from the unprotected stack.
+
+                // net wasm stack: [] -> [ir_return_type]
+                encode_post_appl_calling_conv(
+                    Some(ir::VarType::Any),
+                    ctx.options.wasm_multi_value,
+                    ctx.stackptr,
+                    mutctx.scratch_mut(),
+                    expr_builder,
+                );
+            },
+            |mutctx, expr_builder| {
+                // net wasm stack: [] -> [ir_return_type]
+                encode_post_appl_calling_conv(
+                    Some(ir::VarType::Any),
+                    ctx.options.wasm_multi_value,
+                    ctx.stackptr,
+                    mutctx.scratch_mut(),
+                    expr_builder,
+                );
```

Comment:
Using `encode_post_appl_calling_conv` seems a bit dangerous to me - you shouldn't assume that the "normal" calling convention places Any on the unprotected stack.  I would recomment creating a new function that (unlike `encode_post_appl_calling_conv`) only does the second part  (i.e.
```rust
            // Return value is on the unprotected stack
            // Recall that the unprotected stack grows downward
            // The convention is to put the data at [stackptr-size, stackptr) where `size` is the size of the target type

            // return_ptr = stackptr - size
            // net wasm stack: [] -> [return_ptr]
            expr_builder.global_get(stackptr);
            expr_builder.i32_const(size_in_memory(vartype) as i32);
            expr_builder.i32_sub();

            // load the source value from memory
            // net wasm stack: [return_ptr] -> [source_type]
            encode_load_memory(0, vartype, vartype, scratch, expr_builder);
```
)

Reply:
that makes sense. Will do that